File: Engine/Sources/SorobanEngine/Persistence/DataStore.swift
import Anzan
import Foundation
import SQLite3

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

/// SQLite-backed storage for DATA sheets — imported records at volumes the
/// JSON manifest shouldn't carry. Values are read lazily (indexed lookups /
/// single range queries), so opening a workbook never loads tables into
/// memory. SQLite ships with macOS; this is a deliberately small wrapper.
public final class DataStore {
    public struct TableInfo: Equatable, Sendable {
        public let name: String
        public let rows: Int
        public let columns: Int
    }

    public enum DataStoreError: Error, CustomStringConvertible {
        case sqlite(String)
        public var description: String {
            if case .sqlite(let message) = self { return "data store error: \(message)" }
            return "data store error"
        }
    }

    public let url: URL
    private var db: OpaquePointer?
    private var valueStatement: OpaquePointer?

    public init(url: URL) throws {
        self.url = url
        guard sqlite3_open_v2(url.path, &db,
                              SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, nil) == SQLITE_OK else {
            throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
        }
        try exec("PRAGMA journal_mode=WAL")
        try exec("""
            CREATE TABLE IF NOT EXISTS tables(
                name TEXT PRIMARY KEY COLLATE NOCASE,
                rows INTEGER NOT NULL, cols INTEGER NOT NULL)
            """)
        try exec("""
            CREATE TABLE IF NOT EXISTS cells(
                t TEXT COLLATE NOCASE, r INTEGER, c INTEGER, v TEXT NOT NULL,
                PRIMARY KEY(t, r, c)) WITHOUT ROWID
            """)
    }

    deinit {
        sqlite3_finalize(valueStatement)
        sqlite3_close(db)
    }

    private func exec(_ sql: String) throws {
        guard sqlite3_exec(db, sql, nil, nil, nil) == SQLITE_OK else {
            throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
        }
    }

    // MARK: Tables

    public func tables() throws -> [TableInfo] {
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }
        guard sqlite3_prepare_v2(db, "SELECT name, rows, cols FROM tables ORDER BY name",
                                 -1, &statement, nil) == SQLITE_OK else {
            throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
        }
        var result: [TableInfo] = []
        while sqlite3_step(statement) == SQLITE_ROW {
            result.append(TableInfo(name: String(cString: sqlite3_column_text(statement, 0)),
                                    rows: Int(sqlite3_column_int64(statement, 1)),
                                    columns: Int(sqlite3_column_int64(statement, 2))))
        }
        return result
    }

    public func info(name: String) -> TableInfo? {
        (try? tables())?.first { $0.name.compare(name, options: .caseInsensitive) == .orderedSame }
    }

    /// Imports a rectangular table (one transaction; empty values skipped —
    /// the store is sparse like the grid).
    public func createTable(name: String, rows: [[String]]) throws {
        let columnCount = rows.map(\.count).max() ?? 0
        try exec("BEGIN")
        do {
            var statement: OpaquePointer?
            defer { sqlite3_finalize(statement) }
            guard sqlite3_prepare_v2(
                db, "INSERT OR REPLACE INTO cells(t, r, c, v) VALUES(?, ?, ?, ?)",
                -1, &statement, nil) == SQLITE_OK else {
                throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
            }
            for (r, row) in rows.enumerated() {
                for (c, value) in row.enumerated() where !value.isEmpty {
                    sqlite3_reset(statement)
                    sqlite3_bind_text(statement, 1, name, -1, SQLITE_TRANSIENT)
                    sqlite3_bind_int64(statement, 2, Int64(r))
                    sqlite3_bind_int64(statement, 3, Int64(c))
                    sqlite3_bind_text(statement, 4, value, -1, SQLITE_TRANSIENT)
                    guard sqlite3_step(statement) == SQLITE_DONE else {
                        throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
                    }
                }
            }
            var meta: OpaquePointer?
            defer { sqlite3_finalize(meta) }
            sqlite3_prepare_v2(db, "INSERT OR REPLACE INTO tables(name, rows, cols) VALUES(?, ?, ?)",
                               -1, &meta, nil)
            sqlite3_bind_text(meta, 1, name, -1, SQLITE_TRANSIENT)
            sqlite3_bind_int64(meta, 2, Int64(rows.count))
            sqlite3_bind_int64(meta, 3, Int64(columnCount))
            guard sqlite3_step(meta) == SQLITE_DONE else {
                throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
            }
            try exec("COMMIT")
        } catch {
            try? exec("ROLLBACK")
            throw error
        }
    }

    /// Edits one cell: empty/nil deletes (the store is sparse). The imported
    /// table is the workbook's own copy — edits never touch the source CSV.
    public func setValue(_ value: String?, table: String, row: Int, column: Int) throws {
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }
        if let value, !value.isEmpty {
            guard sqlite3_prepare_v2(
                db, "INSERT OR REPLACE INTO cells(t, r, c, v) VALUES(?, ?, ?, ?)",
                -1, &statement, nil) == SQLITE_OK else {
                throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
            }
            sqlite3_bind_text(statement, 1, table, -1, SQLITE_TRANSIENT)
            sqlite3_bind_int64(statement, 2, Int64(row))
            sqlite3_bind_int64(statement, 3, Int64(column))
            sqlite3_bind_text(statement, 4, value, -1, SQLITE_TRANSIENT)
        } else {
            guard sqlite3_prepare_v2(
                db, "DELETE FROM cells WHERE t = ? AND r = ? AND c = ?",
                -1, &statement, nil) == SQLITE_OK else {
                throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
            }
            sqlite3_bind_text(statement, 1, table, -1, SQLITE_TRANSIENT)
            sqlite3_bind_int64(statement, 2, Int64(row))
            sqlite3_bind_int64(statement, 3, Int64(column))
        }
        guard sqlite3_step(statement) == SQLITE_DONE else {
            throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
        }
    }

    public func dropTable(name: String) throws {
        try exec("BEGIN")
        var cells: OpaquePointer?
        sqlite3_prepare_v2(db, "DELETE FROM cells WHERE t = ?", -1, &cells, nil)
        sqlite3_bind_text(cells, 1, name, -1, SQLITE_TRANSIENT)
        sqlite3_step(cells)
        sqlite3_finalize(cells)
        var meta: OpaquePointer?
        sqlite3_prepare_v2(db, "DELETE FROM tables WHERE name = ?", -1, &meta, nil)
        sqlite3_bind_text(meta, 1, name, -1, SQLITE_TRANSIENT)
        sqlite3_step(meta)
        sqlite3_finalize(meta)
        try exec("COMMIT")
    }

    // MARK: Values (0-based row/column)

    public func value(table: String, row: Int, column: Int) -> String? {
        if valueStatement == nil {
            sqlite3_prepare_v2(db, "SELECT v FROM cells WHERE t = ? AND r = ? AND c = ?",
                               -1, &valueStatement, nil)
        }
        guard let statement = valueStatement else { return nil }
        sqlite3_reset(statement)
        sqlite3_bind_text(statement, 1, table, -1, SQLITE_TRANSIENT)
        sqlite3_bind_int64(statement, 2, Int64(row))
        sqlite3_bind_int64(statement, 3, Int64(column))
        guard sqlite3_step(statement) == SQLITE_ROW,
              let text = sqlite3_column_text(statement, 0) else { return nil }
        return String(cString: text)
    }

    /// All stored values in a rectangle, one query (for range expansion).
    public func values(table: String, rows: ClosedRange<Int>,
                       columns: ClosedRange<Int>) throws -> [(row: Int, column: Int, value: String)] {
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }
        guard sqlite3_prepare_v2(db, """
            SELECT r, c, v FROM cells
            WHERE t = ? AND r BETWEEN ? AND ? AND c BETWEEN ? AND ?
            ORDER BY r, c
            """, -1, &statement, nil) == SQLITE_OK else {
            throw DataStoreError.sqlite(String(cString: sqlite3_errmsg(db)))
        }
        sqlite3_bind_text(statement, 1, table, -1, SQLITE_TRANSIENT)
        sqlite3_bind_int64(statement, 2, Int64(rows.lowerBound))
        sqlite3_bind_int64(statement, 3, Int64(rows.upperBound))
        sqlite3_bind_int64(statement, 4, Int64(columns.lowerBound))
        sqlite3_bind_int64(statement, 5, Int64(columns.upperBound))
        var result: [(Int, Int, String)] = []
        while sqlite3_step(statement) == SQLITE_ROW {
            result.append((Int(sqlite3_column_int64(statement, 0)),
                           Int(sqlite3_column_int64(statement, 1)),
                           String(cString: sqlite3_column_text(statement, 2))))
        }
        return result
    }
}

/// A worksheet backed by a DataStore table: lazily fetched, editable within
/// its OWN row bounds (a data sheet can exceed the grid's 1,000 rows —
/// references like `Sales!C:50000` are valid against its table size).
public final class DataSheet {
    public let table: String
    public let rowCount: Int
    public let columnCount: Int
    private let store: DataStore

    public init?(table: String, store: DataStore) {
        guard let info = store.info(name: table) else { return nil }
        self.table = table
        self.store = store
        self.rowCount = info.rows
        self.columnCount = min(info.columns, Spreadsheet.columnCount)
    }

    /// Raw stored text (UI display / copy). 0-based.
    public func rawValue(row: Int, column: Int) -> String {
        store.value(table: table, row: row, column: column) ?? ""
    }

    /// Edits one cell of the imported copy (0-based; within the table's
    /// rectangle — growing the table is a future feature).
    public func setRawValue(_ value: String, row: Int, column: Int) throws {
        guard row >= 0, row < rowCount, column >= 0, column < columnCount else {
            throw EngineError.domainError(message: "cell is outside this data sheet")
        }
        try store.setValue(value, table: table, row: row, column: column)
    }

    /// Resolver semantics, mirroring the grid: empty → 0, text → error.
    /// Row is 1-based (reference syntax), bounded by the TABLE, not the grid.
    public func numericValue(column: String, row: Int) throws -> BigDecimal {
        guard let columnIndex = CellAddress.columnIndex(forName: column),
              columnIndex < columnCount, (1...max(rowCount, 1)).contains(row) else {
            throw EngineError.domainError(message: "cell \(column):\(row) is outside this data sheet")
        }
        guard let text = store.value(table: table, row: row - 1, column: columnIndex) else {
            return .zero
        }
        guard let value = BigDecimal(string: text) else {
            throw EngineError.domainError(message: "cell \(column):\(row) is not a number")
        }
        return value
    }

    /// Range expansion, grid-consistent: numeric values only, text and empty
    /// skipped. One SQL query regardless of rectangle size.
    public func numericValues(fromColumn: String, fromRow: Int,
                              toColumn: String, toRow: Int) throws -> [BigDecimal] {
        guard let from = CellAddress.columnIndex(forName: fromColumn),
              let to = CellAddress.columnIndex(forName: toColumn),
              fromRow >= 1, toRow >= 1 else {
            throw EngineError.domainError(message: "range is outside this data sheet")
        }
        let rows = min(fromRow, toRow) - 1...max(fromRow, toRow) - 1
        let columns = min(from, to)...max(from, to)
        return try store.values(table: table, rows: rows, columns: columns)
            .compactMap { BigDecimal(string: $0.value) } // text (headers) skipped
    }
}

/// Minimal RFC 4180-style CSV: quoted fields, escaped quotes (""), CR/LF/CRLF.
public enum CSV {
    public static func parse(_ text: String) -> [[String]] {
        var rows: [[String]] = []
        var row: [String] = []
        var field = ""
        var inQuotes = false
        var iterator = text.makeIterator()
        var pending: Character?

        func endField() { row.append(field); field = "" }
        func endRow() {
            endField()
            if !(row.count == 1 && row[0].isEmpty) { rows.append(row) }
            row = []
        }

        while let c = pending ?? iterator.next() {
            pending = nil
            if inQuotes {
                if c == "\"" {
                    if let next = iterator.next() {
                        if next == "\"" { field.append("\"") }   // escaped quote
                        else { inQuotes = false; pending = next } // closing quote
                    } else { inQuotes = false }
                } else {
                    field.append(c)
                }
            } else {
                switch c {
                case "\"" where field.isEmpty: inQuotes = true
                case ",": endField()
                // NB: "\r\n" is ONE Swift Character (grapheme cluster) —
                // match all three line-end forms as single characters.
                case "\n", "\r", "\r\n": endRow()
                default: field.append(c)
                }
            }
        }
        if !field.isEmpty || !row.isEmpty { endRow() }
        return rows
    }

    /// The inverse of `parse`: rows → RFC 4180-style text (\n line ends).
    /// Fields are quoted only when they need it (comma, quote, or newline);
    /// quotes double inside quoted fields. `parse(encode(rows)) == rows`.
    public static func encode(_ rows: [[String]]) -> String {
        rows.map { row in
            row.map(encodeField).joined(separator: ",")
        }.joined(separator: "\n") + (rows.isEmpty ? "" : "\n")
    }

    private static func encodeField(_ field: String) -> String {
        guard field.contains(where: { $0 == "," || $0 == "\"" || $0.isNewline }) else {
            return field
        }
        return "\"" + field.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }
}

File: rust/engine/src/data_store.rs
//! SQLite-backed storage for DATA sheets — the port of
//! `swift/Engine/Sources/SorobanEngine/Persistence/DataStore.swift`:
//! imported records at volumes the JSON manifest shouldn't carry. Values are
//! read lazily (indexed lookups / single range queries), so opening a
//! workbook never loads tables into memory. Swift links the macOS system
//! SQLite; here `rusqlite` (bundled) plays that role — still a deliberately
//! small wrapper.

use crate::cell_address::CellAddress;
use crate::spreadsheet::Spreadsheet;
use anzan::{BigDecimal, EngineError};
use rusqlite::{Connection, OpenFlags};
use std::fmt;
use std::ops::RangeInclusive;
use std::path::{Path, PathBuf};
use std::rc::Rc;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TableInfo {
    pub name: String,
    pub rows: usize,
    pub columns: usize,
}

#[derive(Debug)]
pub enum DataStoreError {
    Sqlite(String),
}

impl fmt::Display for DataStoreError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let DataStoreError::Sqlite(message) = self;
        write!(f, "data store error: {message}")
    }
}

impl std::error::Error for DataStoreError {}

impl From<rusqlite::Error> for DataStoreError {
    fn from(error: rusqlite::Error) -> Self {
        DataStoreError::Sqlite(error.to_string())
    }
}

pub struct DataStore {
    path: PathBuf,
    conn: Connection,
}

impl DataStore {
    pub fn new(path: &Path) -> Result<Self, DataStoreError> {
        let conn = Connection::open_with_flags(
            path,
            OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE,
        )?;
        // `journal_mode` returns a row, so this must be a query, not execute.
        conn.query_row("PRAGMA journal_mode=WAL", [], |_| Ok(()))?;
        conn.execute(
            "CREATE TABLE IF NOT EXISTS tables(
                name TEXT PRIMARY KEY COLLATE NOCASE,
                rows INTEGER NOT NULL, cols INTEGER NOT NULL)",
            [],
        )?;
        conn.execute(
            "CREATE TABLE IF NOT EXISTS cells(
                t TEXT COLLATE NOCASE, r INTEGER, c INTEGER, v TEXT NOT NULL,
                PRIMARY KEY(t, r, c)) WITHOUT ROWID",
            [],
        )?;
        Ok(Self {
            path: path.to_path_buf(),
            conn,
        })
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    /// Flush the write-ahead log into the main database file so a byte copy of
    /// `path` (what the package writer does) captures every committed row.
    /// Without this, recent imports/edits sit in the `-wal` and are lost when
    /// only `data.sqlite` is copied into a `.soroban` package.
    pub fn checkpoint(&self) -> Result<(), DataStoreError> {
        self.conn
            .query_row("PRAGMA wal_checkpoint(TRUNCATE)", [], |_| Ok(()))?;
        Ok(())
    }

    // MARK: Tables

    pub fn tables(&self) -> Result<Vec<TableInfo>, DataStoreError> {
        let mut statement = self
            .conn
            .prepare("SELECT name, rows, cols FROM tables ORDER BY name")?;
        let rows = statement.query_map([], |row| {
            Ok(TableInfo {
                name: row.get(0)?,
                rows: row.get::<_, i64>(1)? as usize,
                columns: row.get::<_, i64>(2)? as usize,
            })
        })?;
        rows.collect::<Result<Vec<_>, _>>().map_err(Into::into)
    }

    pub fn info(&self, name: &str) -> Option<TableInfo> {
        let needle = name.to_lowercase();
        self.tables()
            .ok()?
            .into_iter()
            .find(|info| info.name.to_lowercase() == needle)
    }

    /// Imports a rectangular table (one transaction; empty values skipped —
    /// the store is sparse like the grid).
    pub fn create_table(&self, name: &str, rows: &[Vec<String>]) -> Result<(), DataStoreError> {
        let column_count = rows.iter().map(Vec::len).max().unwrap_or(0);
        self.conn.execute_batch("BEGIN")?;
        let result = (|| -> Result<(), DataStoreError> {
            let mut statement = self
                .conn
                .prepare("INSERT OR REPLACE INTO cells(t, r, c, v) VALUES(?, ?, ?, ?)")?;
            for (r, row) in rows.iter().enumerate() {
                for (c, value) in row.iter().enumerate() {
                    if value.is_empty() {
                        continue;
                    }
                    statement.execute(rusqlite::params![name, r as i64, c as i64, value])?;
                }
            }
            self.conn.execute(
                "INSERT OR REPLACE INTO tables(name, rows, cols) VALUES(?, ?, ?)",
                rusqlite::params![name, rows.len() as i64, column_count as i64],
            )?;
            self.conn.execute_batch("COMMIT")?;
            Ok(())
        })();
        if result.is_err() {
            let _ = self.conn.execute_batch("ROLLBACK");
        }
        result
    }

    /// Edits one cell: empty/None deletes (the store is sparse). The imported
    /// table is the workbook's own copy — edits never touch the source CSV.
    pub fn set_value(
        &self,
        value: Option<&str>,
        table: &str,
        row: usize,
        column: usize,
    ) -> Result<(), DataStoreError> {
        match value {
            Some(value) if !value.is_empty() => {
                self.conn.execute(
                    "INSERT OR REPLACE INTO cells(t, r, c, v) VALUES(?, ?, ?, ?)",
                    rusqlite::params![table, row as i64, column as i64, value],
                )?;
            }
            _ => {
                self.conn.execute(
                    "DELETE FROM cells WHERE t = ? AND r = ? AND c = ?",
                    rusqlite::params![table, row as i64, column as i64],
                )?;
            }
        }
        Ok(())
    }

    pub fn drop_table(&self, name: &str) -> Result<(), DataStoreError> {
        self.conn.execute_batch("BEGIN")?;
        let _ = self
            .conn
            .execute("DELETE FROM cells WHERE t = ?", rusqlite::params![name]);
        let _ = self
            .conn
            .execute("DELETE FROM tables WHERE name = ?", rusqlite::params![name]);
        self.conn.execute_batch("COMMIT")?;
        Ok(())
    }

    // MARK: Values (0-based row/column)

    /// Errors read as absent, like Swift's cached-statement path.
    pub fn value(&self, table: &str, row: usize, column: usize) -> Option<String> {
        let mut statement = self
            .conn
            .prepare_cached("SELECT v FROM cells WHERE t = ? AND r = ? AND c = ?")
            .ok()?;
        statement
            .query_row(rusqlite::params![table, row as i64, column as i64], |r| {
                r.get::<_, String>(0)
            })
            .ok()
    }

    /// All stored values in a rectangle, one query (for range expansion).
    /// Returns `(row, column, value)` triples in row-major order.
    pub fn values(
        &self,
        table: &str,
        rows: RangeInclusive<usize>,
        columns: RangeInclusive<usize>,
    ) -> Result<Vec<(usize, usize, String)>, DataStoreError> {
        let mut statement = self.conn.prepare(
            "SELECT r, c, v FROM cells
             WHERE t = ? AND r BETWEEN ? AND ? AND c BETWEEN ? AND ?
             ORDER BY r, c",
        )?;
        let mapped = statement.query_map(
            rusqlite::params![
                table,
                *rows.start() as i64,
                *rows.end() as i64,
                *columns.start() as i64,
                *columns.end() as i64
            ],
            |row| {
                Ok((
                    row.get::<_, i64>(0)? as usize,
                    row.get::<_, i64>(1)? as usize,
                    row.get::<_, String>(2)?,
                ))
            },
        )?;
        mapped.collect::<Result<Vec<_>, _>>().map_err(Into::into)
    }
}

/// A worksheet backed by a DataStore table: lazily fetched, editable within
/// its OWN row bounds (a data sheet can exceed the grid's 1,000 rows —
/// references like `Sales!C:50000` are valid against its table size).
pub struct DataSheet {
    table: String,
    row_count: usize,
    column_count: usize,
    store: Rc<DataStore>,
}

impl DataSheet {
    pub fn new(table: &str, store: Rc<DataStore>) -> Option<Self> {
        let info = store.info(table)?;
        Some(Self {
            table: table.to_string(),
            row_count: info.rows,
            column_count: info.columns.min(Spreadsheet::COLUMN_COUNT),
            store,
        })
    }

    pub fn table(&self) -> &str {
        &self.table
    }

    /// Bounded by the TABLE, not the grid's row count.
    pub fn row_count(&self) -> usize {
        self.row_count
    }

    /// Capped at the grid's 26 columns.
    pub fn column_count(&self) -> usize {
        self.column_count
    }

    /// Raw stored text (UI display / copy). 0-based.
    pub fn raw_value(&self, row: usize, column: usize) -> String {
        self.store
            .value(&self.table, row, column)
            .unwrap_or_default()
    }

    /// Edits one cell of the imported copy (0-based; within the table's
    /// rectangle — growing the table is a future feature).
    pub fn set_raw_value(&self, value: &str, row: usize, column: usize) -> Result<(), EngineError> {
        if row >= self.row_count || column >= self.column_count {
            return Err(EngineError::domain("cell is outside this data sheet"));
        }
        self.store
            .set_value(Some(value), &self.table, row, column)
            .map_err(|error| EngineError::domain(error.to_string()))
    }

    /// Resolver semantics, mirroring the grid: empty → 0, text → error.
    /// Row is 1-based (reference syntax), bounded by the TABLE, not the grid.
    pub fn numeric_value(&self, column: &str, row: i64) -> Result<BigDecimal, EngineError> {
        let column_index = CellAddress::column_index(column).filter(|&c| c < self.column_count);
        let (Some(column_index), true) = (
            column_index,
            (1..=self.row_count.max(1) as i64).contains(&row),
        ) else {
            return Err(EngineError::domain(format!(
                "cell {column}:{row} is outside this data sheet"
            )));
        };
        let Some(text) = self
            .store
            .value(&self.table, (row - 1) as usize, column_index)
        else {
            return Ok(BigDecimal::zero());
        };
        BigDecimal::parse(&text)
            .ok_or_else(|| EngineError::domain(format!("cell {column}:{row} is not a number")))
    }

    /// Range expansion, grid-consistent: numeric values only, text and empty
    /// skipped. One SQL query regardless of rectangle size.
    pub fn numeric_values(
        &self,
        from_column: &str,
        from_row: i64,
        to_column: &str,
        to_row: i64,
    ) -> Result<Vec<BigDecimal>, EngineError> {
        let (Some(from), Some(to), true) = (
            CellAddress::column_index(from_column),
            CellAddress::column_index(to_column),
            from_row >= 1 && to_row >= 1,
        ) else {
            return Err(EngineError::domain("range is outside this data sheet"));
        };
        let rows = (from_row.min(to_row) - 1) as usize..=(from_row.max(to_row) - 1) as usize;
        let columns = from.min(to)..=from.max(to);
        Ok(self
            .store
            .values(&self.table, rows, columns)
            .map_err(|error| EngineError::domain(error.to_string()))?
            .into_iter()
            .filter_map(|(_, _, value)| BigDecimal::parse(&value)) // text (headers) skipped
            .collect())
    }
}

#[cfg(test)]
mod tests;
